spec/integrations/rebalancing/message_order_with_sync_commit.rb

```diff
@@ -0,0 +1,162 @@
+# frozen_string_literal: true
+
+# Messages should not be reprocessed out of order during a rebalance with
+# additional manual synchronous commits using #mark_as_consumed!.
+
+setup_karafka(allow_errors: %w[consumer.consume.error connection.client.poll.error]) do |config|
+  config.kafka[:'max.poll.interval.ms'] = 10_000
+  config.kafka[:'session.timeout.ms'] = 10_000
+  # Default librdkafka auto commit interval
+  config.kafka[:'auto.commit.interval.ms'] = 5_000
+
+  # Attempt ot batch messages to help trigger the condition for a contested message
+  config.max_wait_time = 10_000
+  config.max_messages = 20
+  config.initial_offset = 'latest'
+end
+
+class Consumer < Karafka::BaseConsumer
+  def consume
+    messages.each do |message|
+      # Wait until we receive a batch with more than 3 messages, where the current one follows
+      # at least 2 other messages in order to avoid false positives and give the second consumer
+      # a chance to reprocess the whole batch.
+      condition_met = ((messages.metadata.first_offset + 2)..(messages.metadata.last_offset - 1)).cover?(message.offset)
+
+      # Wait until some amount of messages has been already processed to spice things up
+      if condition_met && DT.data[:contested_message].empty? && DT.data[:consecutive_messages].count >= 60
+        DT[:contested_message] << message
+        sleep(0.1) until DT.data.key?(:second_closed)
+      end
+
+      DT.data[:consecutive_messages] << [message.payload['key'], message.partition, message.offset]
+
+      # #mark_as_consumed! will raise on the contesed message since by that time
+      # the partition will be revoked
+      return_value = mark_as_consumed!(message)
+      DT[:mark_as_consumed_return_values] << [message.payload, return_value]
+      return if return_value == false
```

Comment:
All of the assertions in this spec showcase the behaviour that is expected, except for this construct here (assertions are commented on at the bottom of the spec).